### PR TITLE
coprocessor: implement functions: period_add, period_diff

### PR DESCRIPTION
**Pull request by Seact**
## What have you changed? (mandatory)

implement builtin function: PeriodAdd and PeriodDiff in coprocesser
[#3275](https://github.com/tikv/tikv/issues/3275)

## What are the type of the changes? (mandatory)
improvement

## How has this PR been tested? (mandatory)
unit test

## Does this PR affect documentation (docs) update? (mandatory)
no

## Does this PR affect tidb-ansible update? (mandatory)
no

## Refer to a related PR or issue link (optional)
issue [#3275](https://github.com/tikv/tikv/issues/3275)
correct mistake at [#4395](https://github.com/tikv/tikv/pull/4395)

### CODE REVIEW

**Comment by rleungx** [comment]
@Seact Hi, what are the differences between this PR and #4396?

**Comment by Seact** [comment]
> 
> 
> @Seact Hi, what are the differences between this PR and #4396?
#4396 does not correspond with the mysql implementation when dealing with i64 overflowing

**Comment by Seact** [comment]
Why "DCO — The sign-off is missing.", although the commit is signed-off ?

**Comment by rleungx** [comment]
@Seact All your commits included in this PR should sign off the DCO. It seems the commit `8309253` is not sign-off.

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -1341,6 +1387,51 @@ mod tests {
...
+                ScalarFuncSig::PeriodAdd,
+                Datum::I64(arg1),
+                Datum::I64(arg2),
+                Datum::I64(exp),
+            );
+        }
+    }
+
+    #[test]
+    fn test_period_diff() {
```
Comment (rleungx) [inline src/coprocessor/dag/expr/builtin_time.rs]: Can you add more tests? You can directly port them from `TiDB`.

**Comment by rleungx** [comment]
@AndreMouche PTAL.

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -322,6 +322,52 @@ impl ScalarFunc {
...
+        }
+    }
+
+    fn month_to_period(month: u64) -> u64 {
+        if month == 0 {
+            return 0;
+        }
+        let year = month / 12;
+        if year < 70 {
+            year * 100 + month % 12 + 1 + 200000 // 2000*100
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: I think `(year+2000)*100 + month%12` will be more understandable

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 3]
```diff
@@ -322,6 +322,52 @@ impl ScalarFunc {
         Ok(None)
     }
 
+    fn period_to_month(period: u64) -> u64 {
+        if period == 0 {
+            return 0;
+        }
+        let (year, month) = (period / 100, period % 100);
+        if year < 70 {
+            year * 12 + month - 1 + 24000 // 2000*12
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: I think `(year+2000)*12+month-1 ` will be more understandable

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 4]
```diff
@@ -322,6 +322,52 @@ impl ScalarFunc {
...
+        } else if year < 100 {
+            year * 12 + month - 1 + 22800 // 1900*12
+        } else {
+            year * 12 + month - 1
+        }
+    }
+
+    fn month_to_period(month: u64) -> u64 {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: why make this function inside `ScalarFunc`?

**Comment by AndreMouche** [comment]
@rleungx @lonng PTAL

**Pull request merged**